File: src/core/cpu/ARM/data_processing.rs
use crate::core::{CPU, Bus, Flag};
use crate::core::utils::is_bit_set;
use crate::core::cpu::CPU_mode;


impl CPU {
    #[inline]
    fn ARM_get_2nd_operand(&mut self, instr: u32) -> (u32, bool) {
        if is_bit_set(instr, 25) {  // immediate
            self.barrel_shifter_rotate_right(instr&0xFF, (instr&0xF00) >> 7)
        } else {  // register
            self.barrel_shifter_operand(instr>>4, self.register[(instr&0xF) as usize])
        }
    }

    #[inline(always)]
    pub fn ARM_get_operands(&mut self, instr: u32) -> (u32, (u32, bool)) {
        if !is_bit_set(instr, 25) && is_bit_set(instr, 4) {
            self.register[15] += 4;
            let ops = (
                self.register[((instr >> 16)&0xF) as usize],
                self.ARM_get_2nd_operand(instr)
            );
            self.register[15] -= 4;
            ops
        } else {
            (
                self.register[((instr >> 16)&0xF) as usize],
                self.ARM_get_2nd_operand(instr)
            )
        }
    }

    #[inline]
    pub fn ARM_AND(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp = op1&op2;
        self.register_write(((instr >> 12)&0xF) as usize, tmp, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_EOR(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp= op1^op2;
        self.register_write(((instr >> 12)&0xF) as usize, tmp, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_SUB(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let (tmp, overflow) = (op1 as i32).overflowing_sub(op2 as i32);
        let d = ((instr >> 12)&0xF) as usize;
        self.register_write(d, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            if d == 15 {
                match self.mode {
                    CPU_mode::sys | CPU_mode::usr => {},
                    CPU_mode::fiq => self.set_mode(CPU_mode::from(self.fiq_spsr)),
                    _ => self.set_mode(CPU_mode::from(self.bank_reg[self.mode as usize][2]))
                }
            }
            self.set_flag(Flag::C, tmp as u32 <= op1);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_RSB(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let (tmp, overflow) = (op2 as i32).overflowing_sub(op1 as i32);
        self.register_write(((instr >> 12)&0xF) as usize, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, tmp as u32 <= op2);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow)
        }
    }

    #[inline]
    pub fn ARM_ADD(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let (tmp, overflow) = (op1 as i32).overflowing_add(op2 as i32);
        self.register_write(((instr >> 12)&0xF) as usize, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, op1.overflowing_add(op2).1);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_ADC(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let c = (self.cpsr&Flag::C != 0) as u64;
        let tmp = op1 as u64+op2 as u64+c;
        self.register_write(((instr >> 12)&0xF) as usize, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, tmp > u32::MAX as u64);
            self.set_flag(Flag::Z, tmp as u32 == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
            self.set_flag(Flag::V, (op1 as i32).overflowing_add((op2+c as u32) as i32).1);
        }
    }

    #[inline]
    pub fn ARM_SBC(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let c = (self.cpsr&Flag::C != 0) as u32;
        let (tmp, overflow) = (op1 as i32).overflowing_sub((op2+1-c) as i32);
        self.register_write(((instr >> 12)&0xF) as usize, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, op2 as u64 - c as u64 + 1 <= op1 as u64);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_RSC(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let c = (self.cpsr&Flag::C != 0) as u32;
        let (tmp, overflow) = (op2 as i32).overflowing_sub((op1+1-c) as i32);
        self.register_write(((instr >> 12)&0xF) as usize, tmp as u32, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, op1 as u64 - c as u64 + 1 <= op2 as u64);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_TST(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp = op1&op2;

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_TEQ(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp = op1^op2;

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_CMP(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let (tmp, overflow) = (op1 as i32).overflowing_sub(op2 as i32);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, tmp as u32 <= op1);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_CMN(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, _)) = self.ARM_get_operands(instr);
        let (tmp, overflow) = (op1 as i32).overflowing_add(op2 as i32);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, op1.overflowing_add(op2).1);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp < 0);
            self.set_flag(Flag::V, overflow);
        }
    }

    #[inline]
    pub fn ARM_ORR(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp= op1|op2;
        self.register_write(((instr >> 12)&0xF) as usize, tmp, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_MOV(&mut self, bus: &mut Bus, instr: u32) {
        let (_, (op2, carry)) = self.ARM_get_operands(instr);
        self.register_write(((instr >> 12)&0xF) as usize, op2, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, op2 == 0);
            self.set_flag(Flag::N, op2&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_BIC(&mut self, bus: &mut Bus, instr: u32) {
        let (op1, (op2, carry)) = self.ARM_get_operands(instr);
        let tmp = op1& !op2;
        self.register_write(((instr >> 12)&0xF) as usize, tmp, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, tmp == 0);
            self.set_flag(Flag::N, tmp&0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_MVN(&mut self, bus: &mut Bus, instr: u32) {
        let (_, (op2, carry)) = self.ARM_get_operands(instr);
        self.register_write(((instr >> 12)&0xF) as usize, !op2, bus);

        if is_bit_set(instr, 20) {
            self.set_flag(Flag::C, carry);
            self.set_flag(Flag::Z, !op2 == 0);
            self.set_flag(Flag::N, !op2 &0x80000000 != 0);
        }
    }

    #[inline]
    pub fn ARM_MRS(&mut self, bus: &mut Bus, instr: u32) {
        let dest = ((instr >> 12)&0xF) as usize;
        let psr = if is_bit_set(instr, 22) {  // current mode spsr
            match self.mode {
                CPU_mode::sys | CPU_mode::usr => self.cpsr, // ???????????? todo: research
                CPU_mode::fiq => self.fiq_spsr,
                _ => self.bank_reg[self.mode as usize][2]
            }
        } else {
            self.cpsr
        };
        
        self.register_write(dest, psr, bus);
    }

    #[inline]
    pub fn ARM_MSR(&mut self, bus: &mut Bus, instr: u32) {
        let src = if is_bit_set(instr, 25) {
            (instr&0xFF).rotate_right((instr&0xF00) >> 7)
        } else {
            self.register[(instr&0xF) as usize]
        };


        let mask = if self.mode == CPU_mode::usr {
            0xF0000000  // in user mode only top 4 bits can change
        } else {
            (is_bit_set(instr, 16) as u32*0xFF)
            | (is_bit_set(instr, 17) as u32*0xFF00)
            | (is_bit_set(instr, 18) as u32*0xFF0000)
            | (is_bit_set(instr, 19) as u32*0xFF000000)
        };

        if is_bit_set(instr, 22) {  // current mode spsr
            let dest = match self.mode {
                CPU_mode::sys | CPU_mode::usr => return,
                CPU_mode::fiq => &mut self.fiq_spsr,
                _ => &mut self.bank_reg[self.mode as usize][2]
            };
            *dest = (*dest& !mask) | (src&mask);
        } else {
            self.cpsr = (self.cpsr& !mask) | (src&mask);
            self.set_mode(CPU_mode::from(self.cpsr));
        }
    }
}

File: core/src/main/kotlin/spAnGB/cpu/arm/DataProcessing.kt
@file:Suppress("NOTHING_TO_INLINE")

package spAnGB.cpu.arm

import spAnGB.cpu.*
import spAnGB.memory.AccessType
import spAnGB.utils.bit
import spAnGB.utils.hex
import spAnGB.utils.toInt
import spAnGB.utils.uLong

val dataProcessingDsl = DataProcessingDsl()

class DataProcessingDsl {
    lateinit var cpu: CPU
    @JvmField
    var instruction: Int = 0
    @JvmField
    var firstOperand: Int = 0
    @JvmField
    var secondOperand: Int = 0

    @JvmField
    var destinationRegister: Int = 0
    @JvmField
    var result: Int = 0

    fun initialize(cpu: CPU) {
        this.cpu = cpu
        this.instruction = cpu.instr

        destinationRegister = (instruction ushr 12) and 0xF
        secondOperand = when {
            !(instruction bit 25) && instruction bit 4 -> {
                cpu.bus.idle()
                cpu.prefetchAccess = AccessType.NonSequential
                cpu.pc += 4
                val tmp = cpu.operand(instruction ushr 4, cpu.registers[instruction and 0xF])
                firstOperand = cpu.registers[(instruction ushr 16) and 0xF]
                cpu.pc -= 4
                tmp
            }
            instruction bit 25 -> {
                firstOperand = cpu.registers[(instruction ushr 16) and 0xF]
                cpu.barrelShifterRotateRight(instruction and 0xFF, (instruction and 0xF00) ushr 7)
            }
            else -> {
                firstOperand = cpu.registers[(instruction ushr 16) and 0xF]
                cpu.operand(instruction ushr 4, cpu.registers[instruction and 0xF])
            }
        }
    }


    inline fun perform(func: DataProcessingDsl.() -> Int) {
        result = func()
        cpu.setRegister(destinationRegister, result)
    }

    inline fun performWithoutDestination(func: DataProcessingDsl.() -> Int) {
        result = func()
        if (destinationRegister == 15) {
            println("[ARM] performWithoutDestination: destination == 15")

            cpu.setCPUMode(CPUMode.byMask[cpu.spsr and 0x1F]!!)
            cpu.state = if (cpu.spsr and CPUFlag.T.mask != 0) CPUState.THUMB else CPUState.ARM
            cpu.cpsr = cpu.spsr
        }
    }

    inline fun setFlags(func: DataProcessingDsl.() -> Unit) {
        if (instruction bit 20) {
            N = result < 0
            Z = result == 0
            func()
        }
    }

    inline fun setFlagsWithShifterCarry(func: DataProcessingDsl.() -> Unit = {}) {
        if (instruction bit 20) {
            N = result < 0
            Z = result == 0
            C = cpu.shifterCarry
            func()
        }
    }

    inline fun checkDestinationPC(): Boolean {
        if (destinationRegister == 15) {
            cpu.apply {
                pc = result
                val s = spsr
                setCPUMode(CPUMode.byMask[s and 0x1F]!!)
                changeState((s and CPUFlag.T.mask != 0).toInt())
                cpsr = s
            }
        }

        return destinationRegister == 15
    }

    inline fun overflow(op2: Int = secondOperand) {
        V = (firstOperand xor op2).inv() and (firstOperand xor result) < 0
    }

    inline fun subOverflow(op1: Int = firstOperand, op2: Int = secondOperand) {
        V = (op1 xor op2) and (op2 xor result).inv() < 0
    }

    inline fun dumbCarry(func: DataProcessingDsl.() -> Long) {  // TODO
        C = func() bit 32
    }

    inline fun dumbBorrow(func: DataProcessingDsl.() -> Long) {  // TODO
        C = !(func() bit 32)
    }

    inline val carry: Int get() = C.toInt()

    inline var N
        get() = cpu[CPUFlag.N]
        set(value) {
            cpu[CPUFlag.N] = value
        }

    inline var Z
        get() = cpu[CPUFlag.Z]
        set(value) {
            cpu[CPUFlag.Z] = value
        }

    inline var C
        get() = cpu[CPUFlag.C]
        set(value) {
            cpu[CPUFlag.C] = value
        }

    inline var V
        get() = cpu[CPUFlag.V]
        set(value) {
            cpu[CPUFlag.V] = value
        }
}


private inline fun instruction(crossinline func: DataProcessingDsl.() -> Unit): CPU.() -> Unit =
    {
        dataProcessingDsl.initialize(this)
        dataProcessingDsl.func()
    }

val armAnd = ARMInstruction(
    { "And" },
    instruction {
        perform { firstOperand and secondOperand }
        setFlagsWithShifterCarry()
    }
)

val armEor = ARMInstruction(
    { "Eor" },
    instruction {
        perform { firstOperand xor secondOperand }
        setFlagsWithShifterCarry()
    }
)

val armTst = ARMInstruction(
    { "Tst" },
    instruction {
        performWithoutDestination { firstOperand and secondOperand }
        setFlagsWithShifterCarry()
    }
)

val armTeq = ARMInstruction(
    { "Teq" },
    instruction {
        performWithoutDestination { firstOperand xor secondOperand }
        setFlagsWithShifterCarry()
    }
)

val armOrr = ARMInstruction(
    { "Orr" },
    instruction {
        perform { firstOperand or secondOperand }
        setFlagsWithShifterCarry()
    }
)

val armMov = ARMInstruction(
    { "Mov" },
    instruction {
        result = secondOperand

        setFlagsWithShifterCarry {
            if (checkDestinationPC()) return@instruction
        }

        cpu.setRegister(destinationRegister,  result)
    }
)

val armBic = ARMInstruction(
    { "Bic" },
    instruction {
        perform { firstOperand and (secondOperand.inv()) }
        setFlagsWithShifterCarry()
    }
)

val armMvn = ARMInstruction(
    { "Mvn" },
    instruction {
        perform { secondOperand.inv() }
        setFlagsWithShifterCarry()
    }
)

val armSub = ARMInstruction(
    { "Sub" },
    instruction {
        result = firstOperand - secondOperand
        setFlags {
            subOverflow()
            dumbBorrow { firstOperand.uLong - secondOperand.uLong }

            if (checkDestinationPC()) return@instruction
        }

        cpu.setRegister(destinationRegister, result)
    }
)

val armRsb = ARMInstruction(
    { "Rsb" },
    instruction {
        perform { secondOperand - firstOperand }
        setFlags {
            subOverflow(secondOperand, firstOperand)
            dumbBorrow { secondOperand.uLong - firstOperand.uLong }
        }
    }
)

val armAdd = ARMInstruction(
    { "Add" },
    instruction {
        perform { firstOperand + secondOperand }
        setFlags {
            overflow()
            dumbCarry { firstOperand.uLong + secondOperand.uLong }
        }
    }
)

val armAdc = ARMInstruction(
    { "Adc" },
    instruction {
        perform { firstOperand + secondOperand + carry }
        setFlags {
            overflow()
            dumbCarry { firstOperand.uLong + secondOperand.uLong + carry.uLong }
        }
    }
)

val armSbc = ARMInstruction(
    { "Sbc" },
    instruction {
        perform { firstOperand - secondOperand - 1 + carry }
        setFlags {
            subOverflow()
            dumbBorrow { firstOperand.uLong - secondOperand.uLong - 1L + carry.uLong }
        }
    }
)

val armRsc = ARMInstruction(
    { "Rsc" },
    instruction {
        perform { secondOperand - firstOperand - 1 + carry }
        setFlags {
            subOverflow(op1 = secondOperand, op2 = firstOperand)
            dumbBorrow { secondOperand.uLong - firstOperand.uLong - 1L + carry.uLong }
        }
    }
)

val armCmp = ARMInstruction(
    { "Cmp" },
    instruction {
        performWithoutDestination { firstOperand - secondOperand }
        setFlags {
            subOverflow()
            dumbBorrow { firstOperand.uLong - secondOperand.uLong }
        }
    }
)

val armCmn = ARMInstruction(
    { "Cmn" },
    instruction {
        performWithoutDestination { firstOperand + secondOperand }
        setFlags {
            overflow()
            dumbCarry { firstOperand.uLong + secondOperand.uLong }
        }
    }
)

val armMrs = ARMInstruction(
    { "Mrs" },
    instruction {
        perform {
            when (instruction bit 22) {
                true -> cpu.spsr
                false -> cpu.cpsr
            }
        }
    }
)

val armMsr = ARMInstruction(
    { "Msr" },
    {
        val src = when (instr bit 25) {
            true -> (instr and 0xFF).rotateRight((instr and 0xF00) ushr 7)
            false -> registers[instr and 0xF]
        }

        val mask: Int = when (mode) {
            CPUMode.User -> 0xF0000000u.toInt()
            else -> (0xFF * (instr bit 16).toInt())
                .or(0xFF00 * (instr bit 17).toInt())
                .or(0xFF0000 * (instr bit 18).toInt())
                .or(0xFF000000u.toInt() * (instr bit 19).toInt())
        }

        when (instr bit 22) {
            true -> {
                spsr = (spsr and mask.inv()) or (src and mask)
            }
            false -> {
                cpsr = (cpsr and mask.inv()) or (src and mask)
                setCPUMode(CPUMode.byMask[cpsr and 0x1F]!!)  // TODO: changing mode by mask?
            }
        }
    }
)